src/shared/completion/use-set-form-completion-mutation.js

```diff
@@ -0,0 +1,94 @@
+import { useAlert, useDataEngine } from '@dhis2/app-runtime'
+import { useQueryClient, useMutation } from '@tanstack/react-query'
+import { useApiAttributeParams } from '../use-api-attribute-params.js'
+import {
+    useDataSetId,
+    useOrgUnitId,
+    usePeriodId,
+} from '../use-context-selection/index.js'
+import { useDataValueSetQueryKey } from '../use-data-value-set/index.js'
+
+const MUTATION_SET_FORM_COMPLETION = {
+    resource: 'dataEntry/dataSetCompletion',
+    type: 'create',
+    data: ({
+        dataSetId,
+        periodId,
+        orgUnitId,
+        categoryComboId,
+        categoryOptionIds,
+        completed,
+    }) => ({
+        dataSet: dataSetId,
+        period: periodId,
+        orgUnit: orgUnitId,
+        attribute: { combo: categoryComboId, options: categoryOptionIds },
+        completed,
+    }),
+}
+
+export default function useSetFormCompletionMutation() {
+    const queryClient = useQueryClient()
+    const engine = useDataEngine()
+    const [dataSetId] = useDataSetId()
+    const [periodId] = usePeriodId()
+    const [orgUnitId] = useOrgUnitId()
+    const {
+        attributeCombo: categoryComboId,
+        attributeOptions: categoryOptionIds,
+    } = useApiAttributeParams()
+
+    const { show: showErrorAlert } = useAlert((message) => message, {
+        critical: true,
+    })
+
+    const mutationFn = ({ completed }) => {
+        const variables = {
+            dataSetId,
+            periodId,
+            orgUnitId,
+            categoryComboId,
+            categoryOptionIds,
+            completed,
+        }
+
+        return engine.mutate(MUTATION_SET_FORM_COMPLETION, { variables })
+    }
+
+    const dataValueSetQueryKey = useDataValueSetQueryKey()
+
+    return useMutation(mutationFn, {
+        retry: 0, // @TODO: Is this correct?
+        mutationKey: dataValueSetQueryKey,
+        onMutate: async ({ variables }) => {
+            // Cancel any outgoing refetches (so they don't overwrite our optimistic delete)
+            await queryClient.cancelQueries(dataValueSetQueryKey)
+
+            // Snapshot the previous value
+            const previousDataValueSet =
+                queryClient.getQueryData(dataValueSetQueryKey)
+
+            // Optimistically overwrite the completion status
+            queryClient.setQueryData(dataValueSetQueryKey, () => ({
+                ...previousDataValueSet,
+                completeStatus: {
+                    ...previousDataValueSet.completeStatus,
```

Comment:
this spread operator seems to leave the lastUpdated and lastUpdatedBy unchanged? Which among other things, leads to the condition in main-tool-bar.js (`data.completeStatus.lastUpdatedBy`) not displaying until the information is refreshed from the backend

Reply:
Yeah, ideally the backend would return the updated values in the response to the POST. But I guess we could create our own object in the meantime? 

If not, it would also be great to fetch this from the backend without the `dataValues`, so we could invalidate just this information, by doing a simple fetch to eg. `dataEntry/dataSetCompletion` (with params for the selection of course).

Reply:
Would you mind if we address this in a separate issue / PR?